src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs

```diff
@@ -287,4 +253,53 @@ private static bool IsWebProject(ProjectResource projectResource)
         var launchProfile = projectResource.GetEffectiveLaunchProfile(throwIfNotFound: true);
         return launchProfile?.ApplicationUrl != null;
     }
+
+    private static void SetAspNetCoreUrls(this IResourceBuilder<ProjectResource> builder)
+    {
+        if (builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            return;
+        }
+
+        builder.WithEnvironment(context =>
+        {
+            if (context.EnvironmentVariables.ContainsKey("ASPNETCORE_URLS"))
+            {
+                // If the user has already set ASPNETCORE_URLS, we don't want to override it.
+                return;
+            }
+
+            var aspnetCoreUrls = new ReferenceExpressionBuilder();
+
+            var processedHttpsPort = false;
+            var first = true;
+
+            static bool IsValidAspNetCoreUrl(EndpointAnnotation e) =>
+                e.UriScheme is "http" or "https" && e.TargetPortEnvironmentVariable is null;
```

Comment:
Why does having a `TargetPortEnvironmentVariable` mean it isn't a valid ASPNETCORE_URLS?

Reply:
It's super indirect. I didn't want to design a new API explicitly opting you into ASP.NET Core URLs, but that's another option.

Reply:
One thing I think may help is to update the docs on WithEndpoint's "env" variable:

> /// <param name="env">An optional name of the environment variable to inject.</param>

It doesn't really describe what is happening here.